src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -382,37 +384,30 @@ public static byte[] UrlEncodeToBytes(byte[] value, int offset, int count)
                 return null;
             }
 
-            int cSpaces = 0;
-            int cUnsafe = 0;
+            bool foundSpaces = false;
+            int unsafeCount = 0;
 
             // count them first
             for (int i = 0; i < count; i++)
             {
                 char ch = (char)value[offset + i];
 
                 if (ch == ' ')
-                    cSpaces++;
+                    foundSpaces = true;
                 else if (!IsUrlSafeChar(ch))
-                    cUnsafe++;
+                    unsafeCount++;
             }
 
             // nothing to expand?
-            if (cSpaces == 0 && cUnsafe == 0)
+            if (!foundSpaces && unsafeCount == 0)
```

Comment:
@hughbe Unfortunately no, I don't think the boilerplate is possible to remove. `stackalloc` allocates variables on the stack as its name suggests, so the moment you `return pChars` the memory referenced by the pointer is deallocated and you are left with a dangling pointer. Same goes for `fixed`: when you leave the `fixed` block, the GC is allowed to move the array around wherever it wants and the pointer could eventually reference invalid memory. Additionally, I don't think C# yet supports generic pointers (although some support for this was added in #7966).

In spite of that, it might be possible to do something like this:

```cs
public static void Alloc<T>(int count, int threshold, Action<void*> action)
{
    count = Unsafe.SizeOf<T>() * count;

    if (count < threshold)
    {
        void* memory = stackalloc byte[count];
        action(memory);
    }
    else
    {
        var bytes = new byte[count];
        fixed (void* memory = bytes)
        {
            action(memory);
        }
    }
}
```

This could introduce additional allocations for the `Action` and closures, though, so all in all I don't think it's possible/practical with the way C# currently supports pointers. Interesting idea, though.